lfu_nots.go

```diff
@@ -0,0 +1,214 @@
+}
+
+// Set sets a new key-value pair
+// Set increments the key usage count too
+// eg:
+// cache.Set("test_key","2")
+// cache.Set("test_key","1")
+// if you try to set a value into same key
+// its usage count will be increased
+// and usage count of "test_key" will be 2 in this example
+func (l *LFUNoTS) Set(key string, value interface{}) error {
+	return l.set(key, value)
+}
+
+// Delete deletes the key and its dependencies
+func (l *LFUNoTS) Delete(key string) error {
+	res, err := l.cache.Get(key)
+	if err != nil && err != ErrNotFound {
+		return err
+	}
+
+	// we dont need to delete if already doesn't exist
+	if err == ErrNotFound {
+		return nil
+	}
+
+	ci := res.(*cacheItem)
+
+	l.remove(ci, ci.freqElement)
+	l.currentSize--
+	return l.cache.Delete(key)
+}
+
+// set sets a new key-value pair
+func (l *LFUNoTS) set(key string, value interface{}) error {
+	res, err := l.cache.Get(key)
+	if err != nil && err != ErrNotFound {
+		return err
+	}
+
+	if err == ErrNotFound {
+		//create new cache item
+		ci := newCacheItem(key, value)
+
+		// if cache size si reached to max size
+		// then first remove lfu item from the list
+		if l.currentSize >= l.size {
+			// then evict some data from head of linked list.
+			l.evict(l.frequencyList.Front())
+		}
+
+		l.cache.Set(key, ci)
+		l.incr(ci)
+
+	} else {
+		//update existing one
+		val := res.(*cacheItem)
+		val.v = value
+		l.cache.Set(key, val)
+		l.incr(res.(*cacheItem))
+	}
+
+	return nil
+}
+
+// entry holds the frequency node informations
+type entry struct {
+	// freqCount holds the frequency number
+	freqCount int
+
+	// itemCount holds the items how many exist in list
+	listEntry map[*cacheItem]byte
+}
+
+// incr increments the usage of cache items
+// incrementing will be used in 'Get' & 'Set' functions
+// whenever these functions are used, usage count of any key
+// will be increased
+func (l *LFUNoTS) incr(ci *cacheItem) {
+	var nextValue int
+	var nextPosition *list.Element
+	// update existing one
+	if ci.freqElement != nil {
+		nextValue = ci.freqElement.Value.(*entry).freqCount + 1
+		// replace the position of frequency element
+		nextPosition = ci.freqElement.Next()
+	} else {
+		// create new frequency element for cache item
+		// ci.freqElement is nil so next value of freq will be 1
+		nextValue = 1
+		// we created new element and its position will be head of linked list
+		nextPosition = l.frequencyList.Front()
+		l.currentSize++
+	}
+
+	// we need to check position first, otherwise it will panic if we try to fetch value of entry
+	if nextPosition == nil || nextPosition.Value.(*entry).freqCount != nextValue {
+		// create new entry node for linked list
+		entry := newEntry(nextValue)
+		if ci.freqElement == nil {
+			nextPosition = l.frequencyList.PushFront(entry)
+		} else {
+			nextPosition = l.frequencyList.InsertAfter(entry, ci.freqElement)
+		}
+	}
+	nextPosition.Value.(*entry).listEntry[ci] = 1
+	ci.freqElement = nextPosition
+
+	if ci.freqElement.Prev() != nil {
```

Comment:
document why do you need to remove item from the list regardless of the list count.